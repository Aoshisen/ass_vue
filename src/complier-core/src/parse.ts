import { NodeTypes } from "./ast";

const enum TagType {
  Start,
  End,
}
export function baseParse(content: string) {
  const context = createContext(content);
  return createRoot(parseChildren(context));
}

function createRoot(children) {
  return { children };
}

function parseChildren(context) {
  const nodes: any = [];
  let node;
  const s = context.source;
  if (s.startsWith("{{")) {
    node = parseInterpolation(context);
  } else if (s[0] === "<") {
    //解析<div></div>
    if (/[a-z]/i.test(s[1])) {
      node = parseElement(context);
    }
  }
  if(node===undefined){
    //解析text类型
    node=parseText(context)
  }
  nodes.push(node);
  return nodes;
}

function parseElement(context) {
  const element = parseTag(context, TagType.Start);
  parseTag(context, TagType.End);
  console.log(context.source, "=============");
  return element;
}

//parseTag 有两个作用，如果是以< 开始的，那么就返回我们的tag 以及type,然后推进我们的context 

//如果是以结尾，那么就推进context就行了，不用返回什么东西
function parseTag(context, type: TagType) {
  //Implement
  // 1.解析tag
  // 2.删除处理完成的代码
  // 正则匹配
  const reg = new RegExp(/^<\/?([a-z]*)/i);
  const match: any = reg.exec(context.source);
  const tag = match[1];
  // console.log(match,"match<<");
  //推进我们解析后的代码

  advanceBy(context, match[0].length);
  //删除我们的左边的右边的闭合尖括号
  advanceBy(context, 1);

  if (type === TagType.End) return;

  return {
    type: NodeTypes.ELEMENT,
    tag,
  };
}

function parseInterpolation(context) {
  //{{message}}
  const openDelimiter = "{{";
  const closeDelimiter = "}}";

  //从{{ 开始找}}
  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  );

  //剔除掉{{
  advanceBy(context, openDelimiter.length);

  const rawContentLength = closeIndex - openDelimiter.length;

  //得到message  截取得到
  const rawContent = parseTextData(context,rawContentLength);

  const content = rawContent.trim();

  //推进 提出掉 message}}
  advanceBy(context, rawContentLength + closeDelimiter.length);

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: content,
    },
  };
}

function advanceBy(context: any, length: number) {
  context.source = context.source.slice(length);
}

function createContext(content: string) {
  return {
    source: content,
  };
}

function parseText(context: any): any {
  //1.获取当前的内容
  //推进字符串
  const content = parseTextData(context,context.source.length);
  // console.log(context.source,"source");
   return {
        type: NodeTypes.TEXT,
        content,
      }
}

function parseTextData(context: any,length) {
  const content = context.source.slice(0, length);
  advanceBy(context, content.length);
  return content;
}

