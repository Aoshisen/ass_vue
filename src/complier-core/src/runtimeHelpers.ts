export const TO_DISPLAY_STRING = Symbol("toDisplayString");
export const CREATE_ELEMENT_VNODE=Symbol("create_element_vnode")

export const helperMapName = {
  [TO_DISPLAY_STRING]: "toDisplayString",
  [CREATE_ELEMENT_VNODE]:"create_element_vnode"
};
