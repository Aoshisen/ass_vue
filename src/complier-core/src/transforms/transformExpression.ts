// ok
import { NodeTypes } from "../ast";

export function transformExpression(node) {
  //复合类型节点不会进来
  if (node.type === NodeTypes.INTERPOLATION) {
    node.content = processExpression(node.content);
  }
}

function processExpression(node) {
  node.content = `_ctx.${node.content}` ;
  return node;
}
