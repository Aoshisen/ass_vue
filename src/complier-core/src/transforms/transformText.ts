import { NodeTypes } from "../ast";

//因为需要延后执行，所以我们返回一个函数
export function transformText(node) {
  if (node.type === NodeTypes.ELEMENT) {
    return () => {
      let currentContainer;
      const { children } = node;
      for (let i = 0; i < children.length; i++) {
        const child = children[i];
        //判断当前的child 是不是普通的 hi 或者是插值类型
        if (isText(child)) {
          //找到下一个
          for (let j = i + 1; j < children.length; j++) {
            const next = children[j];
            if (isText(next)) {
              if (!currentContainer) {
                //初始化
                currentContainer = children[i] = {
                  type: NodeTypes.COMPOUND_EXPRESSION,
                  children: [child],
                };
              }
              currentContainer.children.push(" + ");
              currentContainer.children.push(next);
              children.splice(j, 1);
              //数组删除了数组结构发生了变化，保证指向正确
              j--;
            } else {
              currentContainer = undefined;
              break;
            }
          }
        }
      }
    };
  }
}

function isText(node) {
  return node.type === NodeTypes.INTERPOLATION || node.type === NodeTypes.TEXT;
}
