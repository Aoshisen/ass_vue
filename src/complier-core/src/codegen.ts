import { NodeTypes } from "./ast";
import { CREATE_ELEMENT_VNODE, TO_DISPLAY_STRING, helperMapName } from "./runtimeHelpers";

export function generator(ast) {
  const context = createCodegenContext();
  const { push } = context;
  const args = ["_ctx", "_cache"];

  genFunctionPreamble(ast, context);

  const functionName = "render";
  const signature = args.join(", ");

  push(`function ${functionName}(${signature}){`);
  push("return ");
  genNode(ast.codegenNode, context);
  push("}");
  return {
    code: context.code,
  };
}

function genFunctionPreamble(ast, context) {
  const { push } = context;
  const VueBinging = "Vue";

  // const helpers=["toDisplayString"]
  const aliasHelper = (s) => `${helperMapName[s]}:_${helperMapName[s]}`;
  if (ast.helpers.length) {
    push(`const {${ast.helpers.map(aliasHelper).join(", ")}}=${VueBinging}`);
  }
  push("\n");
  push("return ");
}

function genNode(node: any, context: any) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpress(node, context);
      break;

    case NodeTypes.ELEMENT:
    genElement(node, context);
      break;
    default:
      break;
  }
}

function genElement(node,context){
  const {push ,helper}=context;
  const {tag}=node
  push(`${helper(CREATE_ELEMENT_VNODE)}("${tag}")`)
}

function genText(node, context) {
  const { push } = context;
  push(`'${node.content}'`);
}

function genInterpolation(node, context) {
  const { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);
  push(")");
}

function createCodegenContext() {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
    helper(key) {
      return `_${helperMapName[key]}`;
    },
  };
  return context;
}

function genExpress(node: any, context: any) {
  const { push } = context;
  push(`${node.content}`);
}
// return function render(_ctx,cache){
// return "hi"
// }
