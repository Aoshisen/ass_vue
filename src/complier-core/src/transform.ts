import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING } from "./runtimeHelpers";

export function transform(root: any, options: any={}) {
  const context = createTransformContext(root, options);
  travelNode(root, context);
  createRootCodegen(root);
  root.helpers=[...context.helpers.keys()]
}

//rootCodegen Node for codegen
function createRootCodegen(root:any){
  root.codegenNode=root.children[0]
}

function createTransformContext(root: any, options: any) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers:new Map(),
    helper(key){
      context.helpers.set(key,1)
    }
  };
  return context;
}

function travelNode(node: any, context) {
  const nodeTransforms = context.nodeTransforms;
  for (let i = 0; i < nodeTransforms.length; i++) {
    const nodeTransform = nodeTransforms[i];
    nodeTransform(node,context);
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING)
      break;
      case NodeTypes.ELEMENT:
      case NodeTypes.ROOT:
        travelChildren(node,context);
      break;
    default:
      break;
  }

}

function travelChildren(node: any, context: any) {
  const children = node.children;
    for (let i = 0; i < children.length; i++) {
      const node = children[i];
      travelNode(node, context);
    }
}
