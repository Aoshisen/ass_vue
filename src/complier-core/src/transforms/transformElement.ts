import { NodeTypes } from "../ast";
import { CREATE_ELEMENT_VNODE } from "../runtimeHelpers";


export default function transformElement (node,context) {
if(node.type=== NodeTypes.ELEMENT){
    context.helper(CREATE_ELEMENT_VNODE)
}
}