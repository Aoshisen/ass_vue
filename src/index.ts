// ass-vue 的出口文件
import { baseCompile } from "./complier-core/src";
import * as runtimeDom from "./runtime-dom";
import { registerRuntimeCompiler } from "./runtime-dom";

export * from "./runtime-dom";

// export * from "./reactivity";

export { baseCompile } from "./complier-core/src";

function compileToFunction(template) {
  const { code } = baseCompile(template);
  //这里的code 就是我们代码compile 生成的代码；第一个参数是我们的函数参数
  const render = new Function("Vue", code)(runtimeDom);
  return render;
}

registerRuntimeCompiler(compileToFunction);
