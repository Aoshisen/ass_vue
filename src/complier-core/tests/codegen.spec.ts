import { generator } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import transformElement from "../src/transforms/transformElement";
import { transformExpression } from "../src/transforms/transformExpression";

describe('codeGen', () => {
    it("string",() => { 
        const ast=baseParse("hi");

        transform(ast)

        const {code}=generator(ast)

        expect(code).toMatchSnapshot()
        
     })

     it("interpolation",() => { 
 
        const ast=baseParse("{{message}}");

        transform(ast,{nodeTransforms:[transformExpression]})

        const {code}=generator(ast)

        expect(code).toMatchSnapshot()
      })

    it.only("element",() => { 
      const ast =baseParse("<div></div>")
      transform(ast,{nodeTransforms:[transformElement]})
      const {code}=generator(ast)
      expect(code).toMatchSnapshot()
     })
});