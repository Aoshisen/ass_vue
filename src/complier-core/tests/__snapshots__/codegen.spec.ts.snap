// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codeGen element 1`] = `
"const {create_element_vnode:_create_element_vnode}=Vue
return function render(_ctx, _cache){return _create_element_vnode("div")}"
`;
